### PR TITLE
unfurling: integrate Deliverer with the unfurl machinery CORE-9248

### PR DESCRIPTION
**Pull request by mmaxim**
Patch does the following:

1.) Introduce new class `Unfurler`, whose main job is to run unfurling end-to-end, and to queue up a chat message with the unfurl. Maintain the state of all unfurls, and allow outsiders to know about that state for the purposes of sending the unfurl message.
2.) Refactor `ConversationBackedStorage` to use a `uid` as a parameter instead of pulling from global state.
3.) Integrate `Deliverer` with `Unfurler` in a similar manner as `attachments.Uploader`. That is, we check to make sure unfurling is complete before sending the message. We do not count unfurling in progress as a failure, and it also doesn't block any other messages.

### CODE REVIEW

File: go/chat/unfurl/unfurler.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	u.retryCh = ch
+}
+
+func (u *Unfurler) SetTestingUnfurlCh(ch chan *chat1.Unfurl) {
+	u.unfurlCh = ch
+}
+
+func (u *Unfurler) Complete(ctx context.Context, outboxID chat1.OutboxID) {
+	defer u.Trace(ctx, func() error { return nil }, "Complete(%s)", outboxID)()
+	if err := u.G().GetKVStore().Delete(u.taskKey(outboxID)); err != nil {
```
Comment (joshblum) [inline go/chat/unfurl/unfurler.go]: call `doneUnfurling` here for safety?

File: go/chat/sender.go [hunk 2]
```diff
@@ -1146,6 +1151,47 @@ func (s *Deliverer) processAttachment(ctx context.Context, obr chat1.OutboxRecor
...
+		s.G().Unfurler.Retry(ctx, obr.OutboxID)
+		return obr, errDelivererUnfurlInProgress
+	case types.UnfurlerTaskStatusFailed:
+		s.G().Unfurler.Retry(ctx, obr.OutboxID)
+		return obr, errors.New("failed to unfurl")
+	}
+	return obr, nil
+}
+
+func (s *Deliverer) processSpecialMessage(ctx context.Context, obr chat1.OutboxRecord) (chat1.OutboxRecord, error) {
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/sender.go]: Sure

[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/sender.go]: Sure

File: go/chat/unfurl/unfurler_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	}
+}
+
+type dummyDeliverer struct {
+	types.MessageDeliverer
+}
+
+func (d dummyDeliverer) ForceDeliverLoop(ctx context.Context) {}
+
+func TestUnfurler(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/unfurl/unfurler_test.go]: Sure

File: go/chat/sender.go [hunk 4]
```diff
@@ -1252,14 +1298,18 @@ func (s *Deliverer) deliverLoop() {
 					obr.OutboxID, s.clock.Now().Sub(obr.Ctime.Time()))
 				err = delivererExpireError{}
 			} else {
-				// Check for an attachment message and process based on upload status
-				obr, err = s.processAttachment(bctx, obr)
-				if err == errDelivererUploadInProgress {
+				// Check for special messages and process based on completion status
+				obr, err = s.processSpecialMessage(bctx, obr)
+				switch err {
+				case errDelivererUploadInProgress:
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/sender.go]: I kind of like the separate ones.

**Comment by joshblum** [comment]
The error messages can be different but just of a special “inProgress” type

File: go/chat/unfurl/unfurler.go (hunk 1 above)
Comment (mmaxim) [reply go/chat/unfurl/unfurler.go]: Don't need it.

**Pull request merged**